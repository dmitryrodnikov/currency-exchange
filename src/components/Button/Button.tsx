import * as React from 'react';
import styled from 'styled-components';
import {COLORS} from '../../constants/Theme';

function getColorSelector(activeColor: string, inactiveColor: string) {
    return (props: StyledButtonProps) => props.active ? activeColor : inactiveColor;
}

const StyledButton = styled.div`
    padding: 6px 10px;
    margin: 0 2px;
    border-radius: 6px;
    cursor: pointer;
    background: ${getColorSelector(COLORS.green, 'transparent')}
    color: ${getColorSelector(COLORS.white, 'inherit')}
    
    &:hover {
        background-color: ${getColorSelector(COLORS.green, COLORS.grey)};
    }
`;

export class Button<VALUE> extends React.PureComponent<ButtonProps<VALUE>> {
    public render () {
        const {children, active} = this.props;
        return (
            <StyledButton
                active={active}
                onClick={this.clickHandler}
            >
                {children}
            </StyledButton>
        )
    }

    private clickHandler = () => {
        this.props.onClick(this.props.value);
    }
}

export type ButtonClickHandler<VALUE> = (value: VALUE) => void;

interface ButtonProps<VALUE> extends StyledButtonProps {
    value: VALUE;
    active: boolean;
    children: string;
    onClick: ButtonClickHandler<VALUE>;
}

interface StyledButtonProps {
    active: boolean
}

