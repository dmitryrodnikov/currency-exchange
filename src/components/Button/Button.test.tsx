import React from 'react';
import {shallow, render, mount} from 'enzyme';
import {spy} from 'sinon';
import {Button} from './Button';

const BUTTON_VALUE = 'Test value';

describe('Button', () => {
    it('Shallow render snapshots are equal', () => {
        const wrapper = shallow(
            <Button
                active={true}
                value={BUTTON_VALUE}
                onClick={() => {}}
            >
                Text
            </Button>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('Render snapshots are equal', () => {
        const wrapper = render(
            <Button
                active={true}
                value={BUTTON_VALUE}
                onClick={() => {}}
            >
                Text
            </Button>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('Click handler called once with exact argument', () => {
        const clickHandler = spy();
        const wrapper = mount(
            <Button
                active={true}
                value={BUTTON_VALUE}
                onClick={clickHandler}
            >
                Text
            </Button>
        );

        wrapper.find('div').first().simulate('click');

        expect(clickHandler.calledOnceWithExactly(BUTTON_VALUE)).toBe(true);
    });
});
