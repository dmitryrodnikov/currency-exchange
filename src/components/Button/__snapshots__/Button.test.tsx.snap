// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render snapshots are equal 1`] = `
<div
  class="sc-bdVaJa gMmIAm"
>
  Text
</div>
`;

exports[`Button Shallow render snapshots are equal 1`] = `
<styled.div
  active={true}
  onClick={[Function]}
>
  Text
</styled.div>
`;
